src/pg3.rs

```diff
@@ -0,0 +1,329 @@
+
+#[doc(inline)]
+pub use super::wit::pg3::{Error as PgError, *};
+
+/// A pg error
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Failed to deserialize [`DbValue`]
+    #[error("error value decoding: {0}")]
+    Decode(String),
+    /// Pg query failed with an error
+    #[error(transparent)]
+    PgError(#[from] PgError),
+}
+
+/// A type that can be decoded from the database.
+pub trait Decode: Sized {
+    /// Decode a new value of this type using a [`DbValue`].
+    fn decode(value: &DbValue) -> Result<Self, Error>;
+}
+
+impl<T> Decode for Option<T>
+where
+    T: Decode,
+{
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::DbNull => Ok(None),
+            v => Ok(Some(T::decode(v)?)),
+        }
+    }
+}
+
+impl Decode for bool {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Boolean(boolean) => Ok(*boolean),
+            _ => Err(Error::Decode(format_decode_err("BOOL", value))),
+        }
+    }
+}
+
+impl Decode for i16 {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Int16(n) => Ok(*n),
+            _ => Err(Error::Decode(format_decode_err("SMALLINT", value))),
+        }
+    }
+}
+
+impl Decode for i32 {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Int32(n) => Ok(*n),
+            _ => Err(Error::Decode(format_decode_err("INT", value))),
+        }
+    }
+}
+
+impl Decode for i64 {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Int64(n) => Ok(*n),
+            _ => Err(Error::Decode(format_decode_err("BIGINT", value))),
+        }
+    }
+}
+
+impl Decode for f32 {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Floating32(n) => Ok(*n),
+            _ => Err(Error::Decode(format_decode_err("REAL", value))),
+        }
+    }
+}
+
+impl Decode for f64 {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Floating64(n) => Ok(*n),
+            _ => Err(Error::Decode(format_decode_err("DOUBLE PRECISION", value))),
+        }
+    }
+}
+
+impl Decode for Vec<u8> {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Binary(n) => Ok(n.to_owned()),
+            _ => Err(Error::Decode(format_decode_err("BYTEA", value))),
+        }
+    }
+}
+
+impl Decode for String {
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::Str(s) => Ok(s.to_owned()),
+            _ => Err(Error::Decode(format_decode_err(
+                "CHAR, VARCHAR, TEXT",
+                value,
+            ))),
+        }
+    }
+}
+
+/// Native representation of the WIT postgres Date value.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Date(pub chrono::NaiveDate);
```

Comment:
Is there a reason we need a wrapper struct here rather than decoding straight to `chrono::NaiveDate`?  It feels like the user can't do anything the wrapper except call `.0` on it to extract the NaiveDate: is it possible to cut out the middleman?

Reply:
The ole can't implement traits on foreign types.

Reply:
Wait maybe I gave poor advice here. is it can't implement foreign traits on foreign types?

Reply:
Yeah you can't implement _foreign_ traits on _foreign_ types but you can totally implement _local_ traits on foreign types.

Reply:
🤦 yep yep yep.

Reply:
Indeed, we do exactly that on all the other types - stdlib types still count as foreign for orphan rule purposes.